earth2studio/data/utils.py

```diff
@@ -69,6 +74,65 @@ def fetch_data(
 
     return prep_data_array(xr.concat(da, "lead_time"), device=device)
 
+def build_dataset(
+    source: DataSource,
+    time: TimeArray,
+    variable: VariableArray,
+    dataset: zarr.Array,
+    transform: None = None,
+    apache_beam_options: None = None,
+):
+    """
+    Utility function to fetch data for models and load data on the target device.
+    """
+
+    # Set default options
+    if apache_beam_options is None:
+        logger.info("No Apache Beam options provided, using default options (Single-threaded, local runner)")
+        apache_beam_options = PipelineOptions()
+
+    # Check dataset has correct shape
+    if dataset.shape[0] != len(time) or dataset.shape[1] != len(variable):
+        raise ValueError(
+            f"Dataset shape {dataset.shape} does not match time {len(time)} and variable {len(variable)} shape"
+        )
+
+    # Fetch data on first time step to make sure all datasets in cache are created
+    # TODO: Probably remove this
+    for var in variable:
+        source.fetch_array(time[0], var)
+
+    # Create elements for Apache Beam pipeline
+    elements = [(i, t, variable) for i, t in enumerate(time)]
+
+    # Make fetch data DoFn
+    class FetchDataDoFn(beam.DoFn):
+        def __init__(self, source, dataset, total, report_every=1000, transform=None):
+            self.source = source
+            self.dataset = dataset
+            self.total = total
+            self.report_every = report_every
+            self.transform = transform
+            self.progress = Metrics.counter(self.__class__, 'progress')
```

Comment:
What is this providing? Is it possible to have some tqdm feedback for users? I havent ran this yet